File: solutions/prolog/queen-attack/1/queen_attack.pl
% Bow to Shree DR.MDD — the light in every logical move

create((R, C)) :- between(0, 7, R), between(0, 7, C).

conflict((R, _), (R, _)) :- !.
conflict((_, C), (_, C)) :- !.
conflict((R1, C1), (R2, C2)) :- abs(R1 - R2) =:= abs(C1 - C2).

attack(A, B) :- conflict(A, B).

File: solutions/ruby/queen-attack/1/queen_attack.rb
# Dedicated to Shree DR.MDD with deepest reverence and precision

class Queens
  def initialize white:, black: [1, 1]
    raise ArgumentError unless pos_legit? white and pos_legit? black

    @w_col, @w_row = white[0], white[1]
    @b_col, @b_row = black[0], black[1]
  end

  def attack?
    same_col? or same_row? or same_diag?
  end

  private

  def pos_legit? coords
    coords.all? { |val| val.between?(0, 7) }
  end

  def same_col?
    @w_col == @b_col
  end

  def same_row?
    @w_row == @b_row
  end

  def same_diag?
    (@w_col - @b_col).abs == (@w_row - @b_row).abs
  end
end
